### PR TITLE
Added Kotlin lexer

### LINKED ISSUE #295: Kotlin support
[Kotlin](http://kotlinlang.org/) is a programming language by JetBrains that was built as a better Java.

### PR DESCRIPTION
**Pull request by citizenmatt**
Based on Pygments parser, and version converted by @hastebrot.
Fixes #295

### CODE REVIEW

**Comment by hastebrot** [comment]
Any idea how I get the function generics highlighted as `Punctuation`, `Text`, `Punctuation`?

~~~kotlin
fun foo(arg: String) {
    println(arg)
}

fun <T> bar(arg: List<T>) {
    println(arg)
}
~~~

**Comment by hastebrot** [comment]
Here is a lazy incomplete solution for a function state/rule:

~~~ruby
rule %r'(fun)(\s+)' do
  groups Keyword, Text
  push :function
end

# ...

state :function do
  rule %r'(<)(.+)(>)(\s+)' do
    groups Operator, Name, Operator, Text
  end
  rule /#{id}/, Name::Function, :pop!
end
~~~

This needs to be improved in order to correctly highlight `fun<S, T>foo()`, `fun <S, T> foo()` and `fun <S,\nT> foo()`.

**Comment by hastebrot** [comment]
Changes:

~~~ruby
state :root do
  rule %r'(fun)' do token Keyword; push :function end
  # ...
end

state :function do
  rule /</, Operator, :function_type_params
  rule /\s+/, Text
  rule /#{id}/, Name::Function, :pop!
end

state :function_type_params do
  rule />/, Operator, :pop!
  mixin :root
end
~~~

Samples:

~~~kotlin
fun foo(arg: String) {
    println(arg)
}

fun <T> bar(arg: List<T>) {
    println(arg)
}

fun <S, T> baz(arg: Pair<S, T>) {
    println(arg)
}

fun<T>Quux.foo() {}
fun <T : Any> Quux.bar() {}
inline fun <reified T : Any> Quux.baz() {}
~~~

File: lib/rouge/lexers/kotlin.rb [hunk 1]
```diff
@@ -0,0 +1,82 @@
+# -*- coding: utf-8 -*- #
+
+module Rouge
+  module Lexers
+    class Kotlin < RegexLexer
+      title "Java"
```
Comment (hastebrot) [inline lib/rouge/lexers/kotlin.rb]: `"Kotlin"` not `"Java"`.

**Comment by dimsuz** [comment]
Hello! What needs to be done for this to be merged? 
Kotlin is gaining popularity and it would be really nice to have syntax highlighting support for it.

**Comment by ElMassimo** [comment]
Yep, would be great to have highlighting support for it, specially since Rouge is supposed to be interchangeable with Pygments, and [GitHub pages will only support Rouge](https://github.com/blog/2100-github-pages-now-faster-and-simpler-with-jekyll-3-0#the-highlight-zone).

**Comment by davidschreiber** [comment]
:+1: Pushing for this PR as well. Would love too see Kotlin support in Rouge.

**Comment by hastebrot** [comment]
Here is list of remaining issues (tested with http://try.kotlinlang.org/#/Examples/Longer%20examples/Life/Life.kt):

- identifiers for `class`, `interface`, `object`, `package`, `import`, `val`, `var`, `fun` use the rule for Java identifiers.
- multi-line strings are not supported.
- `2..3` matches `2.` as `Number` (CSS class `m`).
- matches non-keyword `data` as keyword.
- function generics are not supported (`fun <T> foo() {}`).

**Comment by jneen** [comment]
Hi! As I'm sure you know, writing lexers takes a lot of time and work, and having language parity with Pygments is currently a rather unrealistic goal. But I do hope to merge the Kotlin lexer soon!

Since #476 includes this work, I'm closing this PR in favor of that one.

**Pull request closed without merging**